File: src/grammar/visitor_impl.py
# -*- coding: utf-8 -*-

from src.grammar.toy_asmParser import toy_asmParser
from src.grammar.toy_asmVisitor import toy_asmVisitor
from src.grammar.operations import *


class VisitorImpl(toy_asmVisitor):
    ops: list[Op] = []

    def visitNum(self, ctx: toy_asmParser.NumContext):
        return int(ctx.getText())

    def visitReg(self, ctx: toy_asmParser.RegContext):
        return Reg(ctx.getText())

    def visitMem(self, ctx: toy_asmParser.MemContext):
        reg = ctx.children[1].getText()
        if len(ctx.children) == 3:
            offset = 0
        else:
            offset = int(ctx.children[2].getText())
        return Mem(reg, offset)

    def visitMove(self, ctx: toy_asmParser.MoveContext):
        p1 = self.visit(ctx.children[1])
        p2 = self.visit(ctx.children[3])
        return Move(p1, p2)

    def visitAdd(self, ctx: toy_asmParser.AddContext):
        p1 = self.visit(ctx.children[1])
        p2 = self.visit(ctx.children[3])
        return Add(p1, p2)

    def visitSub(self, ctx: toy_asmParser.SubContext):
        p1 = self.visit(ctx.children[1])
        p2 = self.visit(ctx.children[3])
        return Sub(p1, p2)

    def visitMul(self, ctx: toy_asmParser.MulContext):
        p1 = self.visit(ctx.children[1])
        return Mul(p1)

    def visitDiv(self, ctx: toy_asmParser.DivContext):
        p1 = self.visit(ctx.children[1])
        return Div(p1)

    def visitCmp(self, ctx: toy_asmParser.CmpContext):
        p1 = self.visit(ctx.children[1])
        p2 = self.visit(ctx.children[3])
        return Cmp(p1, p2)

    def visitJump(self, ctx: toy_asmParser.JumpContext):
        act = ctx.children[0].getText()
        label = ctx.children[1].getText()
        return Jump(act, label)

    def visitCall(self, ctx: toy_asmParser.CallContext):
        label = ctx.children[1].getText()
        return Call(label)

    def visitRet(self, ctx: toy_asmParser.RetContext):
        return Ret()

    def visitPush_op(self, ctx: toy_asmParser.Push_opContext):
        act = ctx.children[0].getText()
        if len(ctx.children) == 2:
            p1 = self.visit(ctx.children[1])
        else:
            p1 = None
        return Push(act, p1)

    def visitPop_op(self, ctx: toy_asmParser.Pop_opContext):
        act = ctx.children[0].getText()
        if len(ctx.children) == 2:
            p1 = self.visit(ctx.children[1])
        else:
            p1 = None
        return Pop(act, p1)

    def visitInput(self, ctx: toy_asmParser.InputContext):
        p1 = self.visit(ctx.children[1])
        return Input(p1)

    def visitPrint(self, ctx: toy_asmParser.PrintContext):
        p1 = self.visit(ctx.children[1])
        return Print(p1)

    def visitDump(self, ctx: toy_asmParser.DumpContext):
        if len(ctx.children) > 1:
            p1 = self.visit(ctx.children[1])
            n = int(ctx.children[3].getText())
            return Dump(p1, n)
        else:
            return Dump()

    def visitPause(self, ctx: toy_asmParser.PauseContext):
        return Pause()

    def visitHalt(self, ctx: toy_asmParser.HaltContext):
        return Halt()

    def visitNop(self, ctx:toy_asmParser.NopContext):
        return Nop()

    def visitRand(self, ctx: toy_asmParser.RandContext):
        p1 = self.visit(ctx.children[1])
        return Rand(p1)

    def visitOpWithLabel(self, ctx: toy_asmParser.OpWithLabelContext):
        if len(ctx.children) == 2:
            label = ctx.children[0].getText().strip()[:-1]  # 去掉label的冒号
            op = self.visit(ctx.children[1])
        else:
            label = ""
            op = self.visit(ctx.children[0])

        assert isinstance(op, Op)
        op.label = label
        self.ops.append(op)

File: src/web_frontend/src/libs/VisitorImpl.ts
import toy_asmVisitor from './grammar/toy_asmVisitor';
import {
    LineContext, LabelContext, OpContext, NumContext, RegContext, MemContext, OffsetContext, StrContext,
    MovContext, AddContext, SubContext, MulContext, DivContext, CmpContext,
    JumpContext, CallContext, RetContext, PushContext, PopContext,
    InputContext, PrintContext, RandContext, HaltContext, BreakContext
} from './grammar/toy_asmParser';
import {Operand, OperandType} from './Operand';

// Parsed operation data (before full Operation is created)
export interface ParsedOp {
    type: string;
    p1?: Operand | null;
    p2?: Operand | null;
    action?: string | null;
    target?: string | null;
}

export interface ParseResult {
    label: string | null;
    op: ParsedOp | null;
}

export class VisitorImpl extends toy_asmVisitor<any> {
    visitNum = (ctx: NumContext): Operand => {
        const val = parseInt(ctx.getText(), 10);
        return new Operand(OperandType.Imm, "", 0, val);
    };

    visitReg = (ctx: RegContext): Operand => {
        const regName = ctx.getText().toLowerCase();
        return new Operand(OperandType.Reg, regName);
    };

    visitOffset = (ctx: OffsetContext): number => {
        return parseInt(ctx.getText(), 10);
    };

    visitMem = (ctx: MemContext): Operand => {
        const regName = ctx.reg().getText().toLowerCase();
        const offsetCtx = ctx.offset();
        const offset = offsetCtx ? this.visitOffset(offsetCtx) : 0;
        return new Operand(OperandType.Mem, regName, offset);
    };

    visitStr = (ctx: StrContext): Operand => {
        const raw = ctx.getText();
        // Parse escape sequences for the evaluated form
        const evaluated = raw.slice(1, -1)
                             .replace(/\\n/g, '\n')
                             .replace(/\\t/g, '\t')
                             .replace(/\\r/g, '\r')
                             .replace(/\\"/g, '"')
                             .replace(/\\\\/g, '\\');
        return new Operand(OperandType.Str, "", 0, 0, [raw, evaluated]);
    };

    visitMov = (ctx: MovContext): ParsedOp => {
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        const p2 = this.visit(ctx.getChild(3)) as Operand;
        return {type: 'mov', p1, p2};
    };

    visitAdd = (ctx: AddContext): ParsedOp => {
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        const p2 = this.visit(ctx.getChild(3)) as Operand;
        return {type: 'add', p1, p2};
    };

    visitSub = (ctx: SubContext): ParsedOp => {
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        const p2 = this.visit(ctx.getChild(3)) as Operand;
        return {type: 'sub', p1, p2};
    };

    visitMul = (ctx: MulContext): ParsedOp => {
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        return {type: 'mul', p1};
    };

    visitDiv = (ctx: DivContext): ParsedOp => {
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        return {type: 'div', p1};
    };

    visitCmp = (ctx: CmpContext): ParsedOp => {
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        const p2 = this.visit(ctx.getChild(3)) as Operand;
        return {type: 'cmp', p1, p2};
    };

    visitJump = (ctx: JumpContext): ParsedOp => {
        const label = ctx.Label().getText();
        const action = ctx.getChild(0)!.getText().toLowerCase();
        return {type: 'jump', action, target: label};
    };

    visitCall = (ctx: CallContext): ParsedOp => {
        const label = ctx.Label().getText();
        return {type: 'call', target: label};
    };

    visitRet = (_ctx: RetContext): ParsedOp => {
        return {type: 'ret'};
    };

    visitPush = (ctx: PushContext): ParsedOp => {
        const action = ctx.getChild(0)!.getText().toLowerCase();
        if (action === 'pushf') {
            return {type: 'push', action: 'pushf'};
        }
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        return {type: 'push', action: 'push', p1};
    };

    visitPop = (ctx: PopContext): ParsedOp => {
        const action = ctx.getChild(0)!.getText().toLowerCase();
        if (action === 'popf') {
            return {type: 'pop', action: 'popf'};
        }
        if (ctx.getChildCount() === 1) {
            return {type: 'pop', action: 'pop'};
        }
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        return {type: 'pop', action: 'pop', p1};
    };

    visitInput = (ctx: InputContext): ParsedOp => {
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        return {type: 'input', p1};
    };

    visitPrint = (ctx: PrintContext): ParsedOp => {
        const action = ctx.getChild(0)!.getText().toLowerCase() as 'print' | 'println';
        if (ctx.getChildCount() === 1) {
            return {type: 'print', action, p1: null};
        }
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        return {type: 'print', action, p1};
    };

    visitRand = (ctx: RandContext): ParsedOp => {
        const p1 = this.visit(ctx.getChild(1)) as Operand;
        return {type: 'rand', p1};
    };

    visitBreak = (ctx: BreakContext): ParsedOp => {
        return {type: 'break'};
    }

    visitHalt = (_ctx: HaltContext): ParsedOp => {
        return {type: 'halt'};
    };

    visitLine = (ctx: LineContext): ParseResult => {
        const labelCtx = ctx.label();
        const opCtx = ctx.op();

        const label = labelCtx ? labelCtx.Label().getText() : null;

        let op: ParsedOp | null = null;
        if (opCtx) {
            const child = opCtx.getChild(0);
            if (child) {
                op = this.visit(child) as ParsedOp;
            }
        }

        return {label, op};
    };
}
